File: src/02_algorithms/01_sorting/04_quicksort/implementations/ts/quick_sort.test.ts
import { quick_sort, partition } from "./quick_sort";

describe("QuickSort", () => {
  test("Returns correct result", async () => {
    let array = [1, 2, 3, 4];
    quick_sort(array);
    let expected_result = [1, 2, 3, 4];
    expect(array).toEqual(expected_result);
  });
  test("Returns correct result", async () => {
    let array = [2, 1];
    quick_sort(array);
    let expected_result = [1, 2];
    expect(array).toEqual(expected_result);
  });
  test("Returns correct result", async () => {
    let array = [2, 1, 3];
    quick_sort(array);
    let expected_result = [1, 2, 3];
    expect(array).toEqual(expected_result);
  });
  test("Returns correct result", async () => {
    let array = [3, 2, 1];
    quick_sort(array);
    let expected_result = [1, 2, 3];
    expect(array).toEqual(expected_result);
  });
  test("Returns correct result", async () => {
    let array = [5, 4, 3, 2, 1];
    quick_sort(array);
    let expected_result = [1, 2, 3, 4, 5];
    expect(array).toEqual(expected_result);
  });
  test("Returns correct result", async () => {
    let array = [8, 1, 4, 5, 2, 6];
    quick_sort(array);
    let expected_result = [1, 2, 4, 5, 6, 8];
    expect(array).toEqual(expected_result);
  });
  test("Partitions an array", async () => {
    let array = [8, 1, 4, 5, 2, 6];
    let pivotIndex = partition(array, 0, array.length - 1);
    expect(pivotIndex).toEqual(4);
    expect(array).toEqual([1, 4, 5, 2, 6, 8]);
  });
});

File: src/02_algorithms/01_sorting/04_quicksort/implementations/go/quick_sort_test.go
package quickSort

import (
	"reflect"
	"testing"
)

func TestQuickSortPartitionsArray(t *testing.T) {
	data := []int{5, 4, 2, 1, 3}
	pivot_index := Partition(data)
	expectedResult := 2

	if !reflect.DeepEqual(pivot_index, expectedResult) {
		t.Errorf("merge(%v) = %v; want %v", data, pivot_index, expectedResult)
	}
}

func TestQuickSortAlreadySortedArray(t *testing.T) {
	data := []int{1, 2, 3, 4}
	QuickSort(data)
	expectedResult := []int{1, 2, 3, 4}

	if !reflect.DeepEqual(data, expectedResult) {
		t.Errorf("merge(%v) = %v; want %v", data, data, expectedResult)
	}
}

func TestQuickSortReversedArrays(t *testing.T) {
	data := []int{4, 3, 2, 1}
	QuickSort(data)
	expectedResult := []int{1, 2, 3, 4}

	if !reflect.DeepEqual(data, expectedResult) {
		t.Errorf("merge(%v) = %v; want %v", data, data, expectedResult)
	}
}

func TestQuickSortSimpleUnsortedArrays(t *testing.T) {
	data := []int{2, 1, 3}
	QuickSort(data)
	expectedResult := []int{1, 2, 3}

	if !reflect.DeepEqual(data, expectedResult) {
		t.Errorf("merge(%v) = %v; want %v", data, data, expectedResult)
	}
}

func TestQuickSortUnsortedArrays(t *testing.T) {
	data := []int{8, 3, 5, 4, 1, 7, 2, 6}
	QuickSort(data)
	expectedResult := []int{1, 2, 3, 4, 5, 6, 7, 8}

	if !reflect.DeepEqual(data, expectedResult) {
		t.Errorf("merge(%v) = %v; want %v", data, data, expectedResult)
	}
}
